src/librbd/WatchNotifyTypes.cc

```diff
@@ -0,0 +1,347 @@
+  ::encode(NOTIFY_OP_ASYNC_PROGRESS, bl);
+  ::encode(async_request_id, bl);
+  ::encode(offset, bl);
+  ::encode(total, bl);
+}
+
+void AsyncProgressPayload::decode(__u8 version, bufferlist::iterator &iter) {
+  ::decode(async_request_id, iter);
+  ::decode(offset, iter);
+  ::decode(total, iter);
+}
+
+void AsyncProgressPayload::dump(Formatter *f) const {
+  f->dump_string("notify_op", "AsyncProgress");
+  f->open_object_section("async_request_id");
+  async_request_id.dump(f);
+  f->close_section();
+  f->dump_unsigned("offset", offset);
+  f->dump_unsigned("total", total);
+}
+
+void AsyncCompletePayload::encode(bufferlist &bl) const {
+  ::encode(NOTIFY_OP_ASYNC_COMPLETE, bl);
+  ::encode(async_request_id, bl);
+  ::encode(result, bl);
+}
+
+void AsyncCompletePayload::decode(__u8 version, bufferlist::iterator &iter) {
+  ::decode(async_request_id, iter);
+  ::decode(result, iter);
+}
+
+void AsyncCompletePayload::dump(Formatter *f) const {
+  f->dump_string("notify_op", "AsyncComplete");
+  f->open_object_section("async_request_id");
+  async_request_id.dump(f);
+  f->close_section();
+  f->dump_int("result", result);
+}
+
+void FlattenPayload::encode(bufferlist &bl) const {
+  ::encode(NOTIFY_OP_FLATTEN, bl);
+  ::encode(async_request_id, bl);
+}
+
+void FlattenPayload::decode(__u8 version, bufferlist::iterator &iter) {
+  ::decode(async_request_id, iter);
+}
+
+void FlattenPayload::dump(Formatter *f) const {
+  f->dump_string("notify_op", "Flatten");
+  f->open_object_section("async_request_id");
+  async_request_id.dump(f);
+  f->close_section();
+}
+
+void ResizePayload::encode(bufferlist &bl) const {
+  ::encode(NOTIFY_OP_RESIZE, bl);
+  ::encode(size, bl);
+  ::encode(async_request_id, bl);
+}
+
+void ResizePayload::decode(__u8 version, bufferlist::iterator &iter) {
+  ::decode(size, iter);
+  ::decode(async_request_id, iter);
+}
+
+void ResizePayload::dump(Formatter *f) const {
+  f->dump_string("notify_op", "Resize");
+  f->dump_unsigned("size", size);
+  f->open_object_section("async_request_id");
+  async_request_id.dump(f);
+  f->close_section();
+}
+
+void SnapCreatePayload::encode(bufferlist &bl) const {
+  ::encode(NOTIFY_OP_SNAP_CREATE, bl);
+  ::encode(snap_name, bl);
+}
+
+void SnapCreatePayload::decode(__u8 version, bufferlist::iterator &iter) {
+  ::decode(snap_name, iter);
+}
+
+void SnapCreatePayload::dump(Formatter *f) const {
+  f->dump_string("notify_op", "SnapCreate");
+  f->dump_string("snap_name", snap_name);
+}
+
+void UnknownPayload::encode(bufferlist &bl) const {
+  assert(false);
+}
+
+void UnknownPayload::decode(__u8 version, bufferlist::iterator &iter) {
+}
+
+void UnknownPayload::dump(Formatter *f) const {
+}
+
+void NotifyMessage::encode(bufferlist& bl) const {
+  ENCODE_START(2, 1, bl);
+  boost::apply_visitor(EncodePayloadVisitor(bl), payload);
+  ENCODE_FINISH(bl);
+}
+
+void NotifyMessage::decode(bufferlist::iterator& iter) {
+  DECODE_START(1, iter);
+
+  int notify_op;
```

Comment:
missed this before - I think this needs to be a fixed size int type to be cross platform. can be fixed in a separate pr